Add db H2

Difirton opened the pull request.

Difirton committed:
<feat> - change db to

Difirton committed:
<feat> - add JDBC dependency

Difirton committed:
<feat> - all tests pass
<feat> - add test db
<fix> - change diagram
<feat> - add repositories and mappers

Difirton committed:
<feat> - all tests pass
<feat> - add test db
<fix> - change diagram
<feat> - add repositories and mappers

Difirton committed:
<feat> - add StandardCRUDRepository

Difirton committed:
<feat> - optimized method to update film in JdbcFilmRepositoryImpl

Difirton committed:
<refactor> - united in JdbcFilmRepositoryImpl delete and insert film genres in single method

Difirton committed:
<refactor> - SQL moved to separate lines
<feat> - add UserFriendRepositoryLazyMapper
<feat> - implemented in UserFriendRepository methods to hit all and find by id

Difirton committed:
<feat> - add tests

Difirton committed:
<feat> - add documentation in StandardCRUDRepository

Difirton committed:
<feat> - add documentation in custom validator

Frilock commented on src/main/java/ru/yandex/practicum/filmorate/config/mapper/FilmRepositoryLazyMapper.java:
```diff
@@ -0,0 +1,25 @@
+                .id(rs.getLong("id"))
+                .name(rs.getString("name"))
+                .description(rs.getString("description"))
+                .releaseDate(rs.getDate("release_date").toLocalDate())
```
А это notNull поле? Тут NPE не получите?

Difirton replied:
Исправил. Не сообразил, что тут приведение, когда писал.

Frilock commented on src/main/java/ru/yandex/practicum/filmorate/config/mapper/FilmRepositoryLazyMapper.java:
```diff
@@ -0,0 +1,25 @@
+                .duration(rs.getInt("duration"))
+                .rate(rs.getInt("rate"))
+                .ratingMPA(RatingMPA.builder()
+                        .id(rs.getLong("rating_mpa_id")).build()).build();
```
завершающий билд лучше на новой строке

Difirton replied:
Хорошо, поправил, действительно стало симпатичнее выглядеть.

Frilock commented on src/main/java/ru/yandex/practicum/filmorate/config/mapper/GenreRepositoryMapper.java:
```diff
@@ -0,0 +1,17 @@
+public class GenreRepositoryMapper implements RowMapper<Genre> {
+    @Override
+    public Genre mapRow(ResultSet rs, int rowNum) throws SQLException {
+        return Genre.builder().id(rs.getLong("id"))
```
тоже лучше переносы сделайте 
.id
.name
.build

Difirton replied:
Да, я видимо решил тут строчки поэкономить))

Frilock commented on src/main/java/ru/yandex/practicum/filmorate/error/GenreNotFoundException.java:
```diff
@@ -0,0 +1,12 @@
+
+    public GenreNotFoundException(Long id) {
+        super("Genre id not found : " + id);
+        log.warn("Genre id not found : {}", id);
```
Видел у кого то уже нечто похожее, может быть у Вас так в примерах. Подскажите, почему в момент броска исключения в лог пишется warn, а не error? На мой взгляд, это не логично.

Difirton replied:
В обучении, вроде, пример был, потому что я помню что думал какой уровень логирования брать, и решил warn или error, поэтому решил просто посмотреть в материале.

Difirton replied:
В любом случае поправил на error. Вроде действительно логично логгировать ексепшены в error.

Frilock commented on src/main/java/ru/yandex/practicum/filmorate/repository/FilmRepository.java:
```diff
@@ -1,13 +1,12 @@
-    @Query(value = "SELECT * FROM films ORDER BY rate DESC LIMIT ?1", nativeQuery = true)
     List<Film> findPopularFilmsByRate(Integer count);
+
+    int[][] updateAll(List<Film> films);
```
А почему такое странное возвращаемое значение?) Что потом с этими массивами происходить будет/должно?

Difirton replied:
Подумал еще над вариантом убрать, но решил пока оставить. До следующего спринта сделаю небольшую документацию, что возвращают данные методы. 
Просто дальше групповой проект, я думаю там уже можно будет увидеть нужны эти значения или нет, если что сотру и компилятор мне быстро подскажет, где надо убрать return.

Frilock commented on src/main/java/ru/yandex/practicum/filmorate/repository/StandardCRUDRepository.java:
```diff
@@ -0,0 +1,65 @@
+ *
+ * @author Dmitriy Kruglov
+ */
+public interface StandardCRUDRepository<M> {
```
Такой комментарий больше из разряда своего опыта на проектах - обычно все таки используют для типизации букву T, предполагаю что от слова type

Difirton replied:
У меня была мысль, что M - model, хотя сейчас подумал, что у меня в проекте вообще E - entity должно быть.))

Difirton replied:
Еще раз подумал, решил заменить М на Т, пусть будет более понятно остальным.

Frilock commented on src/main/java/ru/yandex/practicum/filmorate/repository/impl/JdbcFilmRepositoryImpl.java:
```diff
@@ -0,0 +1,164 @@
+    private static final String SQL_SELECT_POPULAR_FILMS = "SELECT * FROM films ORDER BY rate DESC LIMIT ?";
+
+    @Autowired
+    public JdbcFilmRepositoryImpl(JdbcOperations jdbcOperations, FilmRepositoryEagerMapper eagerFilmMapper,
```
Не знаю, писал уже или нет, но чтоб не писать такой конструктор можно от lombok использовать RequairedArgsConstructor и поля нормально заинжектятся, но это вкусовщина конечно

Difirton replied:
Не писали, я об этом знаю, но каждый раз забываю, что Spring настолько умен.

Frilock commented on src/main/java/ru/yandex/practicum/filmorate/repository/impl/JdbcFilmRepositoryImpl.java:
```diff
@@ -0,0 +1,164 @@
+        this.jdbcOperations.update(connection -> {
+            PreparedStatement ps = connection.prepareStatement(SQL_INSERT_FILMS_ALL_ARGS,
+                    Statement.RETURN_GENERATED_KEYS);
+            ps.setString(1, film.getName());
```
Можно в отдельный метод map.. вынести

Difirton replied:
Разобрал, код больше не дублируется.

Frilock commented on src/main/java/ru/yandex/practicum/filmorate/repository/impl/JdbcUserFriendRepositoryImpl.java:
```diff
@@ -0,0 +1,112 @@
+    @Override
+    public UserFriend save(User user, User friend, boolean isApproved) {
+        this.jdbcOperations.update(SQL_INSERT_ALL_FIELDS, user.getId(), friend.getId(), isApproved);
+        return UserFriend.builder().user(user).friend(friend).approved(isApproved).build();
```
Обычно тоже в строчку не пишут, когда билдер используют:
user
.friend
,,,

Frilock commented on src/test/java/ru/yandex/practicum/filmorate/config/validator/AfterDateValidatorTest.java:
```diff
@@ -15,33 +16,35 @@
     private final Validator validator;
     @Autowired
-    public AfterDateValidatorTest(Validator validator) {
+    private AfterDateValidatorTest(Validator validator) {
```
А зачем делать приватный конструктор?

Difirton replied:
Мне просто идея сколапсила конструктор в одну строчку и при расстановке  уровней доступа полям, я и конструктору private влепил, не заметив, что это не поле.

Frilock reviewed (changes requested):
Добрый день, в целом как всегда очень хорошая работа. Написал пару комментариев, но в основном это комментарии из разряда "можно лучше" или с вопросами к Вам. В целом на вопросы можете прям тут и ответить в ответе на комментарий.
Единственное, где я бы уточнил:
1. Не получите ли вы NPE в мапперах , всегда ли дата заполняется?
2. Что за странное возвращаемое значение в некоторых методах репозеториев? Зачем возвращать массив int или вообще двумерный массив? Какая в этом логика ? Будет ли значения эти использоваться дальше?

Difirton committed:
<refactoring> - replace constructors to @RequiredArgsConstructor
<feat> - add check to null in mappers
<refactoring> - JdbcFilmRepositoryImpl added methods createPreparedStatement and mapFilmInStatement to remove duplicate code

Difirton committed:
<refactoring> - replace M to T

Frilock reviewed (approved):
Молодцы, принял
